### PR TITLE
HIVE-27185: Iceberg: Cache iceberg table while loading for stats.

### PR DESCRIPTION
**Pull request by ayushtkn**
### What changes were proposed in this pull request?

Cache iceberg table while reading stats.

### Why are the changes needed?

Perf benifits.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing Tests.

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 1]
```diff
@@ -52,12 +52,13 @@ private IcebergTableUtil() {
...
-   * @param hmsTable the HMS table
-   * @param skipCache if set to true there won't be an attempt to retrieve the table from SessionState
+   * @param hmsTable      the HMS table
+   * @param skipCache     if set to true there won't be an attempt to retrieve the table from SessionState
+   * @param suffix        the suffix to use for cache.
    * @return the Iceberg table
    */
   static Table getTable(Configuration configuration, org.apache.hadoop.hive.metastore.api.Table hmsTable,
-      boolean skipCache) {
+      boolean skipCache, String suffix) {
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:61]: Why do we need the suffix?

Comment (ayushtkn) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:61]: We want to cache the table for Stats only, if you fetch once for Stats, then use that till end, if you fetch for something else, we don't use that.

In most cases that won't be a problem, but the way we do CTAS and MV's it did creates problem. It first creates an iceberg table and then does the load into that, so in that cases caching would have stale table. Got a test failure as well for those MV cases in the last run

Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:61]: what do you mean only for stats, we don't have an unlimited heap. if you know that stats is stale in certain cases (CTAS, MV) - invalidate the cache.

Comment (ayushtkn) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: I changed the approach, if it is write operation where the rows affected gets changed, I read from the Cache, if not do the same what we do today.

Give a check, if that sounds cool, or if you feel if it that can also cause some issues,

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 2]
```diff
@@ -315,8 +316,9 @@ public boolean canProvideBasicStatistics() {
   @Override
   public Map<String, String> getBasicStatistics(Partish partish) {
     org.apache.hadoop.hive.ql.metadata.Table hmsTable = partish.getTable();
-    TableDesc tableDesc = Utilities.getTableDesc(hmsTable);
-    Table table = Catalogs.loadTable(conf, tableDesc.getProperties());
+    Table table;
+    // For write queries where rows got modified, don't fetch from cache as values could have changed.
+    table = getTable(hmsTable);
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:321]: can we inline the declaration and assignment?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 3]
```diff
@@ -349,6 +351,17 @@ public Map<String, String> getBasicStatistics(Partish partish) {
     return stats;
   }
 
+  private Table getTable(org.apache.hadoop.hive.ql.metadata.Table hmsTable) {
+    Table table;
+    final Optional<QueryState> queryState = SessionStateUtil.getQueryState(conf);
+    if (!queryState.isPresent() || queryState.get().getNumModifiedRows() != 0) {
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:357]: `queryState.get().getNumModifiedRows() > 0`

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionStateUtil.java [hunk 4]
```diff
@@ -115,7 +115,7 @@ public static boolean addCommitInfo(Configuration conf, String tableName, String
     return addResource(conf, COMMIT_INFO_PREFIX + tableName, newCommitInfoMap);
   }
 
-  private static Optional<QueryState> getQueryState(Configuration conf) {
+  public static Optional<QueryState> getQueryState(Configuration conf) {
     return Optional.ofNullable(SessionState.get())
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionStateUtil.java:119]: could we refactor to:

**Review by deniskuzZ — COMMENTED** [review]
LGTM, minor comments

**Review by deniskuzZ — APPROVED** [review]
+1, pending tests

**Pull request merged**